deltachat-ffi/deltachat.h

```diff
@@ -2630,6 +2636,106 @@ char* dc_get_last_error (dc_context_t* context);
 void dc_str_unref (char* str);
 
 
+/**
+ * @class dc_backup_provider_t
+ *
+ * Set up another device.
+ */
+
+/**
+ * Creates an object for sending a backup to another device.
+ *
+ * The backup is sent to through a peer-to-peer channel which is bootstrapped
+ * by a QR-code.  The backup contains the entire state of the account
+ * including credentials.  This can be used to setup a new device.
+ *
+ * Once this function returns the backup is being offered to remove devices.
+ * To wait until one device transferred the backup, use
+ * dc_backup_provider_done().  Alternatively abort the operation using
+ * dc_stop_ongoing_process().
+ *
+ * During execution of the job #DC_EVENT_IMEX_PROGRESS is sent out to indicate
+ * state and progress.
+ *
+ * @memberof dc_backup_sender_t
+ * @param context The context.
+ * @return Opaque object for sending the backup.
+ *    On errors, NULL is returned and dc_get_last_error()returns an error that
+ *    should be shown to the user.
+ */
+dc_backup_provider_t* dc_provide_backup (dc_context_t* context);
+
+
+/**
+ * Returns the QR code text that will offer the backup to other devices.
+ *
+ * The QR code contains a ticket which will validate the backup and provide
+ * authentication for both the provider and the recipient.
+ *
+ * The scanning device should call the scanned text to dc_check_qr().  If
+ * dc_check_qr() returns DC_QR_BACKUP the backup transfer can be started using
+ * dc_get_backup().
+ *
+ * @memberof dc_backup_provider_t
+ * @param context The context.
+ * @param backup_provider The backup provider object as created by
+ *    dc_provide_backup().
+ * @return The text that should be put in the QR code.
+ *    On errors and empty QR code is returned, NULL is never returned.
+ *    the returned string must be released using dc_str_unref() after usage.
+ */
+char* dc_backup_provider_qr (dc_context_t* context, const dc_backup_provider_t* backup_provider);
+
+
+/**
+ * Returns the QR code SVG image that will offer the backup to other devices.
+ *
+ * This works like dc_backup_provider_qr() but returns the text of a rendered
+ * SVG image containing the QR code.
+ *
+ * @memberof dc_backup_provider_t
+ * @param context The context.
+ * @param backup_provider The backup provider object as created by
+ *    dc_provide_backup().
+ * @return The text that should be put in the QR code.
+ *    On errors and empty QR code is returned, NULL is never returned.
+ *    the returned string must be released using dc_str_unref() after usage.
+ */
+char * dc_backup_provider_qr_svg (dc_context_t* context, const dc_backup_provider_t* backup_provider);
+
+/**
+ * Waits for the sending to finish and frees the backup provider object.
+ *
+ * @memberof dc_backup_sender_t
+ * @param context The context.
```

Comment:
It's only used for logging in case waiting fails, because this function is `void`.  Also, it's consistent.  The `dc_backup_provider_qr` also doesn't need a context yet takes one.  I thought it was the usual thing for the APIs to keep taking this Context argument.

Reply:
> I thought it was the usual thing for the APIs to keep taking this Context argument.

In general or in deltachat-ffi? If the latter, no, almost only `@memberof dc_context_t` functions take a `context` parameter. The e.g. `@memberof dc_msg_t` functions only take a `dc_msg_t* msg` and then do
```rust
    let ffi_msg = &*msg;
    let ctx = &*ffi_msg.context;
```

That being said, for me personally either way would be OK, it's not important to me.

Reply:
Yeah, you're right.  This would be introducing a new pattern in the ffi so let's not do that.  I've pushed changes for that now, if you could review those again that'd be appreciated.